### PR TITLE
feat: scan ballot cards one at a time

### PR DESCRIPTION
**Pull request by eventualbuddha**
This required a lot of restructuring. The scanner now scans one card at a time, yielding two paths to scanned image files. It uses the `--batch-prompt` option to `scanimage` which waits for a newline to be sent to `stdin` before scanning the next page (or writing the buffered back side of an already-scanned to a file).

There was a problem using `--batch-prompt` and `--batch-print` in combination with `--swskip`, which is that skipped images do not count toward the requested ballot pages. Therefore, if one side of a ballot card was skipped then the next sheet would be pulled into the feeder and the front side of that sheet would be given as the back side of the first document. To work around this, we no longer use `--swskip`.  Instead, we detect blank pages ourselves using the same threshold of 0.5% foreground (dark) pixels.

`Scanner`'s API changed to have a `scanSheet` method that returns an async generator, i.e. a generator that yields promises. This async generator had to be constructed manually (as opposed to using `async function*`) because of the mismatch between the push-based nature of IO streams and the pull-based nature of generators. To reconcile the two, I created a `DeferredQueue` class, which is an async queue that matches up the consumer and producer sides using the deferred promise pattern.

### CODE REVIEW

File: src/importer.test.ts [hunk 1]
```diff
@@ -38,20 +42,31 @@ test('doImport calls scanner.scanInto', async () => {
...
-  scannerMock.scanInto.mockRejectedValueOnce(new Error('scanner is a banana'))
-  await expect(importer.doImport()).rejects.toThrow(
-    'problem scanning: scanner is a banana'
-  )
+  scanner.scanSheets.mockImplementationOnce(async function* (): AsyncGenerator<
+    Sheet
+  > {
+    yield Promise.reject(new Error('scanner is a banana'))
+  })
+  await expect(importer.doImport()).rejects.toThrow('scanner is a banana')
```
Comment (benadida) [inline src/importer.test.ts:50]: that scanner really needs to not be a banana.

File: src/importer.ts [hunk 2]
```diff
@@ -423,24 +425,21 @@ export default class SystemImporter implements Importer {
...
-        },
-      })
+      debug('scanning starting for batch: %d', batchId)
+      for await (const sheet of this.scanner.scanSheets(
+        this.ballotImagesPath
+      )) {
+        debug('got a ballot card: %o', sheet)
+        await this.cardAdded(sheet, batchId)
+      }
+      debug('scanning completed successfully')
```
Comment (benadida) [inline src/importer.ts:435]: this is beautiful.

File: src/scanner.ts [hunk 3]
```diff
@@ -39,69 +37,118 @@ export enum ScannerImageFormat {
...
       this.format,
       `scanimage ${args.map((arg) => `'${arg}'`).join(' ')}`
     )
 
-    const onFileScannedPromises: (void | Promise<void>)[] = []
-
-    await new Promise((resolve, reject) => {
-      const scanimage = streamExecFile('scanimage', args)
+    const scannedFiles: string[] = []
+    const results = queue<Promise<IteratorResult<Sheet>>>()
```
Comment (benadida) [inline src/scanner.ts:63]: OMG look at that type structure!! ;)

File: src/scanner.ts [hunk 4]
```diff
@@ -39,69 +37,118 @@ export enum ScannerImageFormat {
...
-      new StreamLines(scanimage.stdout!).on('line', (line: string) => {
-        const path = line.trim()
-        debug(
-          'scanimage [pid=%d] reported a scanned file: %s',
-          scanimage.pid,
-          path
-        )
-        onFileScannedPromises.push(onFileScanned?.(path))
-      })
+    scanimage.once('exit', (code) => {
```
Comment (benadida) [inline src/scanner.ts:92]: TIL about `.once`

File: src/scanner.ts (hunk 3 above)
Comment (eventualbuddha) [reply src/scanner.ts:63]: I know, right? Rarely do you get three-level nesting of generics.

File: src/importer.ts (hunk 2 above)
Comment (eventualbuddha) [reply src/importer.ts:435]: And so much code to make this possible. *sigh*

**Pull request merged**